### PR TITLE
feat: adding prom-client metrics + basic metrics (MAPCO-3345)

### PR DESCRIPTION
**Pull request by ronenkapelian**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✔                                                                      |

was added implementation for prom-client libary:

1. added required endpoint (/metrics)
2. adding 3 base metrics:
  a. counter for success ingestion jobs creation
  b. histogram for timings of successfull ingestions
  c. separate histogram for update\new and 
      merger\splitter types
3. include todo remarks to deal with tasker (splitter + merging) on future after code refactoring
4. helm + config adjustment to be supported on openshift deployment

### CODE REVIEW

File: config/default.json [hunk 1]
```diff
@@ -9,6 +9,10 @@
     "logger": {
       "level": "info",
       "prettyPrint": false
+    },
+    "metrics": {
+      "enabled": true,
```
Comment (asafMasa) [inline config/default.json:14]: check that enabled false works

File: config/default.json [hunk 2]
```diff
@@ -9,6 +9,10 @@
     "logger": {
       "level": "info",
       "prettyPrint": false
+    },
+    "metrics": {
```
Comment (asafMasa) [inline config/default.json:13]: please make sure all values in "custom-environment-variables.json" exists in "default.json" file

File: config/default.json [hunk 3]
```diff
@@ -61,5 +65,8 @@
   "ingestionTilesSplittingTiles": {
     "bboxSizeTiles": 10000,
     "tasksBatchSize": 5
+  },
+  "app": {
+    "projectName": "discrete-overseer-creator"
```
Comment (asafMasa) [inline config/default.json]: shouldn't it be in "telemetry" section?

File: config/default.json [hunk 4]
```diff
@@ -9,6 +9,10 @@
     "logger": {
       "level": "info",
       "prettyPrint": false
+    },
+    "metrics": {
+      "enabled": true,
+      "buckets": [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 15, 50, 250, 500]
```
Comment (asafMasa) [inline config/default.json:15]: Shouldn't it be in helm values?

File: src/containerConfig.ts [hunk 5]
```diff
@@ -2,13 +2,16 @@ import config from 'config';
 import { getOtelMixin } from '@map-colonies/telemetry';
 import { trace } from '@opentelemetry/api';
 import { DependencyContainer } from 'tsyringe/dist/typings/types';
```
Comment (asafMasa) [inline src/containerConfig.ts:4]: check if this dependecy is needed.

File: src/containerConfig.ts [hunk 6]
```diff
@@ -30,6 +33,21 @@ export const registerExternalValues = (options?: RegisterOptions): DependencyCon
...
     { token: TOC_ROUTER_SYMBOL, provider: { useFactory: tocRouterFactory } },
+    {
+      token: SERVICES.METRICS_REGISTRY,
+      provider: {
+        useFactory: instancePerContainerCachingFactory((container) => {
+          const config = container.resolve<IConfig>(SERVICES.CONFIG);
+
+          if (config.get<boolean>('telemetry.metrics.enabled')) {
+            client.register.setDefaultLabels({
+              project: config.get('app.projectName'),
```
Comment (asafMasa) [inline src/containerConfig.ts]: why don't you use SERVICE_NAME?

File: src/layers/models/layersManager.ts [hunk 7]
```diff
@@ -109,10 +168,28 @@ export class LayersManager {
...
       } else {
+        // const fetchTimerEnd = this.createTilesSplitterTasksHistogram?.startTimer({ requestType: 'CreateLayer', jobType, taskType });
         const layerZoomRanges = this.zoomLevelCalculator.createLayerZoomRanges(data.metadata.maxResolutionDeg as number);
         jobId = await this.splitTilesTasker.createSplitTilesTasks(data, layerRelativePath, layerZoomRanges, jobType, taskType);
+        // if (fetchTimerEnd) {
+        //   fetchTimerEnd();
+        // }
+      }
+
+      if (fetchTimerEnd) {
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts]: move to after the else {} closer and remove duplication

File: src/layers/models/layersManager.ts [hunk 8]
```diff
@@ -39,14 +49,59 @@ export class LayersManager {
...
+      //   help: 'create merge tiles tasks array calculation duration',
+      //   buckets: config.get<number[]>('telemetry.metrics.buckets'),
+      //   labelNames: ['requestType', 'jobType', 'taskType'] as const,
+      //   registers: [registry],
+      // });
+    }
   }
 
   public async createLayer(data: IngestionParams, overseerUrl: string): Promise<void> {
+    let successCreatingJobTask = true; // if create job task metric completed successfully
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts]: not needed

File: src/layers/models/layersManager.ts [hunk 9]
```diff
@@ -69,6 +124,10 @@ export class LayersManager {
 
     const jobType = await this.getJobType(data);
     const taskType = this.getTaskType(jobType, files, originDirectory);
+
+    const fetchTimerTotalJobsEnd = this.createJobTasksTotalHistogram?.startTimer({ requestType: 'CreateLayer' });
+    const fetchTimerEnd = this.createJobTasksHistogram?.startTimer({ requestType: 'CreateLayer', jobType, taskType });
+
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:98]: try catch as discussed

File: src/layers/models/layersManager.ts [hunk 10]
```diff
@@ -109,10 +168,28 @@ export class LayersManager {
       const layerRelativePath = `${id}/${displayPath}`;
 
       if (taskType === TaskAction.MERGE_TILES) {
+        // const fetchTimerEnd = this.createMergeTilesTasksHistogram?.startTimer({ requestType: 'CreateLayer', jobType, taskType });
         jobId = await this.mergeTilesTasker.createMergeTilesTasks(data, layerRelativePath, taskType, jobType, this.grids, extent, overseerUrl, true);
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts]: try catch on this method, catch will throw
only for success
same histogram for new and update (of tiles merge tasker)

File: src/layers/models/layersManager.ts [hunk 11]
```diff
@@ -109,10 +168,28 @@ export class LayersManager {
...
+        // if (fetchTimerEnd) {
+        //   fetchTimerEnd();
+        // }
       } else {
+        // const fetchTimerEnd = this.createTilesSplitterTasksHistogram?.startTimer({ requestType: 'CreateLayer', jobType, taskType });
         const layerZoomRanges = this.zoomLevelCalculator.createLayerZoomRanges(data.metadata.maxResolutionDeg as number);
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts]: try catch on this method, catch will throw
only for success

File: src/layers/models/layersManager.ts (hunk 10 above)
Comment (asafMasa) [reply src/layers/models/layersManager.ts]: Check if can be aggregated by zoom and square kilometer(km^2) - for splitter also

**Review by asafMasa — CHANGES_REQUESTED** [review]
Good job 👍

File: config/default.json (hunk 2 above)
Comment (ronenkapelian) [reply config/default.json:13]: done

File: config/default.json (hunk 1 above)
Comment (ronenkapelian) [reply config/default.json:14]: done - should be validate after merging - evaluation on openshift with helm

File: config/default.json (hunk 4 above)
Comment (ronenkapelian) [reply config/default.json:15]: added

File: src/containerConfig.ts (hunk 5 above)
Comment (ronenkapelian) [reply src/containerConfig.ts:4]: it's in use for typing ( checked also with npm run lint: fix that usually marks unused importing)

File: src/containerConfig.ts (hunk 6 above)
Comment (ronenkapelian) [reply src/containerConfig.ts]: replaced

File: src/layers/models/layersManager.ts (hunk 8 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts]: removed

File: src/layers/models/layersManager.ts (hunk 9 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts:98]: done

File: src/layers/models/layersManager.ts (hunk 10 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts]: we discuess about it, should consider design and future implemntation that suitable for it. it too much data for metrics purpose

File: src/layers/models/layersManager.ts (hunk 11 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts]: done

File: src/layers/models/layersManager.ts (hunk 7 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts]: done

File: src/containerConfig.ts [hunk 12]
```diff
@@ -30,6 +33,21 @@ export const registerExternalValues = (options?: RegisterOptions): DependencyCon
     { token: LAYERS_ROUTER_SYMBOL, provider: { useFactory: layersRouterFactory } },
     { token: JOBS_ROUTER_SYMBOL, provider: { useFactory: jobsRouterFactory } },
     { token: TOC_ROUTER_SYMBOL, provider: { useFactory: tocRouterFactory } },
+    {
+      token: SERVICES.METRICS_REGISTRY,
+      provider: {
+        useFactory: instancePerContainerCachingFactory((container) => {
```
Comment (asafMasa) [inline src/containerConfig.ts]: Why did you use "instancePerContainerCachingFactory" and not "instanceCachingFactory"?

File: src/layers/models/layersManager.ts [hunk 13]
```diff
@@ -81,116 +118,147 @@ export class LayersManager {
...
+          jobType: jobType,
+          taskType: taskType,
+          msg: `Successfully created job type: ${jobType} and tasks type: ${taskType}`,
+        });
 
-      if (taskType === TaskAction.MERGE_TILES) {
-        jobId = await this.mergeTilesTasker.createMergeTilesTasks(data, layerRelativePath, taskType, jobType, this.grids, extent, overseerUrl, true);
+        if (fetchTimerTotalJobsEnd) {
+          // will add new histogram metrics to count creation duration as total metric
+          fetchTimerTotalJobsEnd();
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts]: Why dont you close it after the catch? (or in the catch if you wan failures?)

File: src/layers/models/layersManager.ts [hunk 14]
```diff
@@ -69,6 +102,10 @@ export class LayersManager {
 
     const jobType = await this.getJobType(data);
     const taskType = this.getTaskType(jobType, files, originDirectory);
+
+    const fetchTimerTotalJobsEnd = this.createJobTasksTotalHistogram?.startTimer({ requestType: 'CreateLayer' });
+    const fetchTimerEnd = this.createJobTasksHistogram?.startTimer({ requestType: 'CreateLayer', jobType, taskType });
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts]: the "fetchTimerEnd" isn't self explanatory

File: src/layers/models/layersManager.ts [hunk 15]
```diff
@@ -69,6 +102,10 @@ export class LayersManager {
 
     const jobType = await this.getJobType(data);
     const taskType = this.getTaskType(jobType, files, originDirectory);
+
+    const fetchTimerTotalJobsEnd = this.createJobTasksTotalHistogram?.startTimer({ requestType: 'CreateLayer' });
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts]: as "fetchTimerTotalJobsEnd" and "fetchTimerEnd" measure the same time, why do we split between them?

**Review by asafMasa — CHANGES_REQUESTED** [review]
😎 looking good. please mind my changes

File: src/layers/models/layersManager.ts (hunk 15 above)
Comment (asafMasa) [reply src/layers/models/layersManager.ts]: Add dashboard json config under assets

**Review by asafMasa — APPROVED** [review]

**Pull request merged**